[DeviceMesh] Simplifying internal bookkeeping with CuTe layout

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161106
* #161016
* #162690

We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta

fduwjj committed:
[DeviceMesh] Simplifying internal bookkeeping with CuTe layout

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Simplifying internal bookkeeping with CuTe layout"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Simplifying internal bookkeeping with CuTe layout"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Simplifying internal bookkeeping with CuTe layout"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to implement the 


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
1. Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout)
2. Separate backend from the mesh operations so that we eventually can let users do lots of operations without initializing any backend.


Concretely, in this PR, we do the following:
1. Replaced all index/offset and its mappings with CuTe Layout and a backend class which handles all the bookkeeping and create backend if needed. Use CuTe layout for both slicing and _flatten.
2. We also started to make devicemesh more functional (first from the backend perspective). Each newly created device mesh is like a universe, all devicemesh transformed out from it (slicing, flatten, unflatten, etc) will share the same backend (PG) while creating a new device mesh will be different universe. So we changed our unit tests accordingly as well.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
1. Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout)
2. Separate backend from the mesh operations so that we eventually can let users do lots of operations without initializing any backend.


Concretely, in this PR, we do the following:
1. Replaced all index/offset and its mappings with CuTe Layout and a backend class which handles all the bookkeeping and create backend if needed. Use CuTe layout for both slicing and _flatten.
2. We also started to make devicemesh more functional (first from the backend perspective). Each newly created device mesh is like a universe, all devicemesh transformed out from it (slicing, flatten, unflatten, etc) will share the same backend (PG) while creating a new device mesh will be different universe. So we changed our unit tests accordingly as well.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fegin reviewed (commented):
With this implementation, `root_to_flatten_mapping` appears to be removed, which simplifies the flatten logic. Is this primarily due to the adoption of layout, or is this due to the fact that this PR already introduce some functional concept that makes it unnecessary?

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
1. Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout)
2. Separate backend from the mesh operations so that we eventually can let users do lots of operations without initializing any backend.


Concretely, in this PR, we do the following:
1. Replaced all index/offset and its mappings with CuTe Layout and a backend class which handles all the bookkeeping and create backend if needed. Use CuTe layout for both slicing and _flatten.
2. We also started to make devicemesh more functional (first from the backend perspective). Each newly created device mesh is like a universe, all devicemesh transformed out from it (slicing, flatten, unflatten, etc) will share the same backend (PG) while creating a new device mesh will be different universe. So we changed our unit tests accordingly as well.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
1. Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout)
2. Separate backend from the mesh operations so that we eventually can let users do lots of operations without initializing any backend.


Concretely, in this PR, we do the following:
1. Replaced all index/offset and its mappings with CuTe Layout and a backend class which handles all the bookkeeping and create backend if needed. Use CuTe layout for both slicing and _flatten.
2. We also started to make devicemesh more functional (first from the backend perspective). Each newly created device mesh is like a universe, all devicemesh transformed out from it (slicing, flatten, unflatten, etc) will share the same backend (PG) while creating a new device mesh will be different universe. So we changed our unit tests accordingly as well.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
1. Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout)
2. Separate backend from the mesh operations so that we eventually can let users do lots of operations without initializing any backend.


Concretely, in this PR, we do the following:
1. Replaced all index/offset and its mappings with CuTe Layout and a backend class which handles all the bookkeeping and create backend if needed. Use CuTe layout for both slicing and _flatten.
2. We also started to make devicemesh more functional (first from the backend perspective). Each newly created device mesh is like a universe, all devicemesh transformed out from it (slicing, flatten, unflatten, etc) will share the same backend (PG) while creating a new device mesh will be different universe. So we changed our unit tests accordingly as well.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
1. Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout)
2. Separate backend from the mesh operations so that we eventually can let users do lots of operations without initializing any backend.


Concretely, in this PR, we do the following:
1. Replaced all index/offset and its mappings with CuTe Layout and a backend class which handles all the bookkeeping and create backend if needed. Use CuTe layout for both slicing and _flatten.
2. We also started to make devicemesh more functional (first from the backend perspective). Each newly created device mesh is like a universe, all devicemesh transformed out from it (slicing, flatten, unflatten, etc) will share the same backend (PG) while creating a new device mesh will be different universe. So we changed our unit tests accordingly as well.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                - Because process group itself is not serializable, we only store the group name in the mapping
+            """
+            # Every mesh_dim_name will only map to one and only one layout.
+            # When the layout has backend initiated, we want to ensure that
```
- only map to one and only -- remove the first 'only'
- intiated -> initialized
- finish the sentence

fduwjj replied:
opps I should have removed the second line.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                - If a name is provided, it will be mapped to the layout for future reference
+                - For world-size layouts with no overrides, it will try to use the default process group
+                - For other layouts, it creates appropriate subgroups using either split_group (more efficient and it only
+                available for nccl communicators) or new_group. The former sometimes is called "eager init" which means
```
'split_group' is not really called 'eager init' is it?  i think you're mixing up terms here.

fduwjj replied:
I want to say it needs to used with eager init.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+            names_to_layouts_map = _mesh_resources.names_to_layouts.setdefault(
+                root_mesh, {}
+            )
+            # Let mesh_dim_name maps to the current layout whose backend has not been initialized.
```
" Let mesh_dim_name maps to the current layout whose backend has not been initialized." is confusing. delete it? (or, what did you mean by this?)

fduwjj replied:
opps might be rebase error..

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+            # When user explicitly pass in a process group, we directly reuse that PG as backend rather
+            # than creating a new one.
+            if group is not None:
+                layouts_to_groups_map[layout] = group.group_name
```
we should validate that this group ranks matches layout first

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+            default_group = _get_default_group()
+
+            if layout == _Layout(((get_world_size(), 1),)) and backend_override == (
+                None,
```
you assume the default group has default backend settings here, but you don't know that.  if someone initialized the default backend unusually, but wants the default settings here, then what?

wconstab replied:
an idea (for future PR) - maybe trying to infer whether the default_pg is a match is a bad idea. 
1) if users init the world group before creating a mesh, they can pass the default_group into the mesh init fn.  This is more explicit, and cleaner.  If they do not, we can just create a new world group.

This proposal would be less convenient but cleaner/explicit. we could discuss with folks if its better than the implicit but imprecise option we have currently.

fduwjj replied:
This is the same behavior as https://github.com/pytorch/pytorch/pull/159371/files#diff-dbbed99b01763453143e50565b636cb37f8f693aefaf18b57d621781114ed1b7.

fduwjj replied:
OK I record your second comment in a doc so that we don't lose tracking it.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                ranks = list(range(get_world_size()))
+                group = (
+                    new_group(
+                        backend="cpu:gloo,cuda:nccl",
```
fine to keep the existing behavior in this PR, but i want to understand the reasoning for creating it with dual-backend.  maybe we want to avoid this in the future?

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                None,
+                None,
+            ),
+            array_mesh: Optional[torch.Tensor] = None,
```
undocumented option. used for what?

fduwjj replied:
Will add it.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                )
+            else:
+                # Generate the pg_ranks_by_dim for pg_creation from the layout.
+                pg_ranks_by_dim = array_mesh if array_mesh is not None else layout.layout_to_global_ranks(get_world_size())
```
would be helpful for the reader if you specify the typing of this clearly.  i have no idea what kind of data to expect in this variable, so its hard to understand the rest of the function

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                backend, pg_options = backend_override
+
+                # If we have a 2D mesh with mesh_dim_names ("dp", "tp"), the group description
+                # of the subgroups would be `mesh_dim_dp` and `mesh_name_tp`.
```
typo, `mesh_name_tp`?  should be `mesh_tp` right? or is there a bug in the actual group_desc code below?

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                    )
+                ):
+                    group = split_group(
+                        parent_pg=default_group,
```
this seems to imply that we're only allowed to perform splits off of the default group.  is there a reason for this? 
(this can be something to address in the future, i suspect this was pre-existing behavior)

wconstab replied:
you misunderstood me. 

If i create a mesh with (DP, TP) and then later I split DP into (DP_shard, DP_replicate)- should I be allowed to use comm_split for this split?  Does it depend on whether DP group was created with new group in the first place, or split from default group in the first place?

fduwjj replied:
this is a great point, let me add a TODO here, I literally just copied the existing code here. (https://github.com/pytorch/pytorch/blob/3bfa35d62e7c5df2b651a3d19b47db74ef284788/torch/distributed/device_mesh.py#L539C13-L539C26)

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                # If bound_device_id exists, it means the nccl communicator has been eagerly initialized
+                # so that we can use `split_group` to create subgroups through `ncclCommSplit`.
+                # In this case, we only need to make one API call (`split_group``) for the subgroup creation
+                # for each mesh dimension. In a 2 * 4 mesh, we only need to make 2 API calls per ranks to create
```
where is the for loop that does these 2 or 6 api calls?

fduwjj replied:
Calling this function has a for loop that is the for-loop for 2 and in the else there is another for-loop... I just blindly copied the original comments.. I think the comment here has wrong math. let me change it.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
+                        group_desc=group_desc,
+                    )
+                else:
+                    # If the subgroup has been already created through `split_group`, we simply loop over `pg_ranks_by_dim`
```
i don't understand this comment. you could perform split more than once, right? so why does 'subgroup was created through split' imply that we need to use new_group below?

fduwjj replied:
Opps this comment is stale, let me change it.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +325,162 @@ def __init__(
                     rank_coords[0].tolist() if rank_coords.size(0) > 0 else None
                 )
 
+        def _maybe_create_backend(
```
as a refactoring thing, it might be nice to extract out the 'create backend' part of this as a separate util, so this function is focused on caching logic.

also i'd rename it to `get_or_create` not `maybe_create`

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -538,132 +531,93 @@ def _setup_world_group_and_device(self):
+        ) -> "DeviceMesh":
+            """
+            Creates a DeviceMesh from an existing backend and layouts. This will create a new device mesh
+            from a same backend universe or transformed from (e.g, flatten, etc.). Although CuTe layout makes
```
from a same backend universe or transformed from - this is both not grammatical and introduces terms that aren't well defined (backend universe =?)

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -538,132 +531,93 @@ def _setup_world_group_and_device(self):
+            # Extract sizes and strides from layouts
+            size_l, stride_l, mesh_size = [], [], []
+            for layout in layouts:
+                size_l.extend(layout.sizes)
```
i was expecting you to use layout algeba directly for this.  would 'compose' do the right thing here?  it doesn't seem like a good sign for 'layout' being the right tool for the job, if we have to convert the layout back to a tensor and iterate it that way for this operation..

fduwjj replied:
Switching to nested layout solves this problem.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -785,10 +750,30 @@ def __getitem__(
-                    submesh = _mesh_resources.create_sub_mesh(
-                        self, mesh_dim_names, slice_mesh_dims
+                    # When users sliced dim_names outside from current mesh, we will check whether
+                    # there is layout overlap. Eventually we will just directly throw error here because
```
what does layout overlap mean? the idea is, if you tried to slice ('dp', and 'dp_tp') this would error?
can't you express this nicely in terms of layout algebra too?

fduwjj replied:
> if you tried to slice ('dp', and 'dp_tp') this would error?

correct that is the purpose.

> can't you express this nicely in terms of layout algebra too?

This and your question in L600 are great points, I thought about it but I didn't find anything we can leverage at this moment. Originally I just want to quickly finish the refactor and check its feasibility and we do see some more complicated picture here now. Let me look further into CuTe and see how we can also leverage layout algebra for this and L600.

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -785,10 +750,30 @@ def __getitem__(
+                            )
+                    res_mesh = DeviceMesh._from_layouts(
+                        self.device_type,
+                        tuple(layouts_sliced),
```
i mentioned this in the following PR, but having a 'type' for this would be nice.

fduwjj replied:
This one is not needed now.

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj commented:
Two more things to add:
1. Add cache for the mesh list when it is not CuTe representable
2. Decide whether to update root bookkeeping inside `_get_or_create_backend` or not. Or we only want to update it in `_flatten`? But we might still need to keep check of root mesh inside `_get_or_create_backend`.

cc: @fegin

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

wconstab commented on torch/distributed/device_mesh.py:
```diff
@@ -488,6 +341,194 @@ def __init__(
+                # Generate the pg_ranks_by_dim for pg_creation from the layout.
+                # When Mesh ranks are not representible by CuTe layout, we want to directly
+                # use the mesh ranks to create PGs.
+                pg_ranks_by_dim = (
```
based on this comment https://github.com/pytorch/pytorch/pull/161106#issuecomment-3224801940
i think we should normalize the way we get ranks for new_group calls.  We should have one way of doing it that we always use, and it should be to take the mesh_tensor and apply the layout to it.

The mesh_tensor would default to `range(n)` if the user did not provide a mesh_tensor, but if they do provide one it lets them fully customize the device ordering.

Wdyt?

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -1026,16 +1095,58 @@ def _flatten(
+                    reverse=True,
+                )
+            )
+            flattened_layout = _Layout(tuple(res_sizes), tuple(res_strides)).coalesce()
```
I think this is the sole direct use of the three support layout functions (coalesce/composition/complement); the others are only done indirectly from local/global ranks

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

ezyang commented:
Unsubbing until you say it's ready

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced `root_to_flatten_mapping` and `flatten_name_to_root_dims` with `layouts_to_groups` and `names_to_layouts`. Basically one layout maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Removed `create_sub_mesh`, `create_flatten_mesh` and `_get_slice_mesh_dims`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.
6. Remove a hidden private attribute `_dim_group_names ` which records all pg_names in a list for a given DeviceMesh, this is not needed since we now have layouts and `layouts_to_groups`. Also use the private flag `_init_backend` for internal bookkeeping whether we initialize backend or not.
7. Replaced `_flatten_mesh_list` with `_layouts` since now `_layouts` can also represent a mesh. 

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

ezyang commented on torch/distributed/_mesh_layout.py:
```diff
@@ -207,3 +207,14 @@ def global_ranks(self, world_size: int) -> list[list[int]]:
+        """
+        previous_span = 1
+        for size, stride in sorted(self.sizes_and_strides, key=lambda x: x[1]):
+            if size * stride <= previous_span:
```
Consider (2, 2):(2, 1). Here, we see this would return True as not (2*1 <= 1) and not (2*2 <= 2). So this function returns True if there is NO overlap, and therefore is misnamed.

Second, consider (1):(1). Here, you will return False as 1*1 <= 1. But this layout does NOT have overlap.

fduwjj replied:
oh after a second think, maybe we also need one more condition which is previous_size >= size. I was thinking about the case when (2, 4):(2, 1) this is an overlap case.

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -156,20 +108,18 @@ def create_flatten_mesh(
+            if device_mesh.ndim == 1 and mesh_dim_name in not_none(
+                device_mesh.mesh_dim_names
+            ):
+                return device_mesh
```
Do you have to check that `mesh_dim_name` is consistent with the name on the device mesh you directly return here? Otherwise this seems to break the invariant that returned mesh has the name passed by `mesh_dim_name`

fduwjj replied:
In this case, because ndim is 1, mesh_dim_names is also just one string list, so  `mesh_dim_name` in device_mesh.mesh_dim_names is sufficient.

ezyang replied:
I'm asking something slightly different. I know that it's valid to use the name from mesh_dim_names. But what if the user requests a different name, that is NOT the one in that list? Now that seems like trouble.

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -156,20 +108,18 @@ def create_flatten_mesh(
-                *self.flatten_name_to_root_dims[root_mesh].keys(),
-            )
+            self.flatten_name_to_root_layout.setdefault(root_mesh, {})
+            invalid_dim_names = not_none(root_mesh.mesh_dim_names)
```
Why do you get to eliminate `flatten_name_to_root_dims`?

fduwjj replied:
great questions, previously when do slicing after flatten, we need to combine dims from `flatten_name_to_root_dims` and dims from mesh to build the submesh. Now we directly store the layout, this is not needed?

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -178,42 +128,43 @@ def create_flatten_mesh(
-            ).reshape(-1, flattened_mesh_dim_size)
+                if (
+                    device_mesh._layout.coalesce()
+                    == self.flatten_name_to_root_layout[root_mesh][mesh_dim_name]
```
Are the meshes here guaranteed to be coalesced? (I can figure this out with more code reading)

fduwjj replied:
Great question again. The answer is no. Because if you slicing contiguous dims yes if not no. For example for the mesh (2,2,2) ("dp", "cp", "tp") when you do mesh["cp", "tp"]._flatten() then yes, but if you do mesh["dp", "tp"]._flatten() then no. But the mesh tensor needs to be flattened, that's why we have the change in L155.

Commit:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -392,6 +172,219 @@ def _get_device_handle(device_type: str = "cuda"):
+            _DeviceMeshBackend acts as a class which handles all the mechanical bookkeepings
+            for the backend initialization of the DeviceMesh and all the following transformations.
+
+            Also we only use a singleton for one DeviceMesh universe, which means all the DeviceMesh
```
The PG per mesh universe will be another PR.

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Simplifying internal bookkeeping with CuTe layout"


We want to refactor the internal bookkeeping of DeviceMesh so that:
Simply the bookkeeping logics and make it generic enough so that it is easy to support new transformations like flatten noncontiguous dim, reshape and unflatten. (We leveraged the CuTe layout). This new layout also let us handle non-contiguous slicing, flatten, transpose possible.


Concretely, in this PR, we do the following:
1. Use the _Layout to handle all index operations rather use a map to record mesh dims.
2. Replaced`flatten_name_to_root_dims` with `flatten_name_to_root_layout`. Basically one (size, stride) pair maps to one PG. One mesh_dim_name can only map to only layout. (More than one mesh_dim_name can map to the same layout).
3. Replaced `_get_slice_mesh_dims` with `_get_slice_mesh_layout`.
4. Use a new function `_maybe_create_backend` to create PG based on mesh_dim_name and layout.
5. Use a new function `_from_layouts` to create a new DeviceMesh from a given layout list.

The PR looks big indeed but we don't change any existing behavior of DeviceMesh, so it is a pure refactor.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta 

[ghstack-poisoned]

fduwjj commented:
Continue in https://github.com/pytorch/pytorch/pull/163213 and will merge that one since this one messed up with EasyCLA

ezyang commented on torch/distributed/device_mesh.py:
```diff
@@ -178,42 +128,43 @@ def create_flatten_mesh(
+                        f"Flatten mesh with mesh_dim_name {mesh_dim_name} has been created before, "
+                        f"but the layout is different from the current layout. "
+                        f"Please specify another valid mesh_dim_name."
+                    )
```
Yes, separate PR please thanks

fduwjj replied:
ok I will separate this part so that it will not be part of refactoring.
